src/CorePolicy.cc

```diff
@@ -0,0 +1,247 @@
+    for (uint32_t i = 1; i < threadClassCoreMap[defaultClass]->numFilled; i++) {
+        uint32_t coreId = entry->map[i];
+        if (Arachne::occupiedAndCount[coreId]->load().numOccupied < minLoaded) {
+            minLoaded = Arachne::occupiedAndCount[coreId]->load().numOccupied;
+            minCoreId = coreId;
+        }
+    }
+    return minCoreId;
+}
+
+/*
+ * Update threadClassCoreMap when a new core is added.  Callers must guarantee
+ * that this function is not called multiple times simultaneously, or at the
+ * same time as removeCore.
+ *
+ * This function will also bootstrap the coreLoadEstimator as soon as it has
+ * a core.
+ *
+ * \param coreId
+ *     The coreId of the new core.
+ */
+void CorePolicy::addCore(int coreId) {
+  CoreList* entry = threadClassCoreMap[defaultClass];
+  entry->map[entry->numFilled] = coreId;
+  entry->numFilled++;
+  if (!loadEstimatorRunning && !Arachne::disableLoadEstimation) {
+    loadEstimatorRunning = true;
+    runLoadEstimator();
+  }
+}
+
+/*
+ * Update threadClassCoreMap when a new core is removed.  Callers must
+ * guarantee that this function is not called multiple times simultaneously,
+ * or at the same time as addCore.
+ *
+ * \param coreId
+ *     The coreId of the doomed core.
+ */
+void CorePolicy::removeCore(int coreId) {
+  CoreList* entry = threadClassCoreMap[defaultClass];
+  entry->numFilled--;
+  int saveCoreId = entry->map[entry->numFilled];
+  for (uint32_t i = 0; i < entry->numFilled; i++) {
+    if (entry->map[i] == coreId) {
+      entry->map[i] = saveCoreId;
+      return;
+    }
+  }
+}
+
+/*
+ * Return the CoreList for a particular threadClass.  This function is not
+ * synchronized with addCore and removeCore and can return stale data.
+ * Callers of this function must ensure that use of stale data is safe.
+ *
+ * \param threadClass
+ *     The threadClass whose CoreList will be returned.
+ */
+CoreList* CorePolicy::getCoreList(ThreadClass threadClass) {
+  return threadClassCoreMap[threadClass];
+}
+
+/*
+ * Bootstrap the core load estimator thread.  Called from addCore.
+ */
+void CorePolicy::runLoadEstimator() {
+    Arachne::createThread(defaultClass, coreLoadEstimator, this);
+}
+
+/*
+ * Periodically wake up and observe the current load in Arachne to determine
+ * whether it is necessary to increase or reduce the number of cores used by
+ * Arachne.  Runs as the top-level method in an Arachne thread.
+ *
+ * \param corePolicy
+ *     Pointer to the CorePolicy which created this thread.
+ */
+void coreLoadEstimator(CorePolicy* corePolicy) {
+    utilizationThresholds = new double[Arachne::maxNumCores];
+    Arachne::PerfStats previousStats;
+    Arachne::PerfStats::collectStats(&previousStats);
+
+    // Each loop cycle makes measurements and uses them to evaluate whether
+    // to increment the core count, decrement it, or do nothing.
+    for (Arachne::PerfStats currentStats; ; previousStats = currentStats) {
+        Arachne::sleep(MEASUREMENT_PERIOD);

```

Comment:
Is there some way of cleaning up so that this method will exit? If not, there should be. Let's discuss.